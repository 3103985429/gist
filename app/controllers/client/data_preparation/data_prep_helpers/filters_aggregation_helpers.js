const winston = require('winston');

const {
    activeSearch_matchOp_orErrDescription,
    activeFilter_matchOp_orErrDescription_fromMultiFilter,
} = require('../../func');

module.exports = {
    addFiltersToQuery,
    addSearchToQuery,
};

/**
 * Helper to concat two arrays with the first being the original aggregation Operators
 * so that the correct order of the operators is not lost.
 * @param {Array<*>} aggregationOperators
 * @param {Array<*>} operatorsToAppend
 * @returns {Array<*>}
 */
const concatWithAggOperators = (aggregationOperators, operatorsToAppend) => aggregationOperators.concat(operatorsToAppend);

/**
 * Uses func.activeSearch_matchOp_orErrDescription to add the matchOps for a search.
 * Concats it to the given aggregationOperators and returns that or throws an error.
 * @param {Object} dataSourceDescription
 * @param {Object} options
 * @param {Object[]=} aggregationOperators
 * @returns {Object[]}
 * @throws {Error}
 */
function addSearchToQuery(dataSourceDescription, options, aggregationOperators = []) {
    // If the search is active, then add it to the query
    if (!options.isSearchActive) {
        return aggregationOperators;
    }

    const { err, matchOps } = activeSearch_matchOp_orErrDescription(dataSourceDescription, options.searchCol, options.searchQ);

    if (err) {
        throw Error(`Search not added to aggregation operators: ${err.message}`);
    }

    winston.debug('Search is active, adding to query');

    return concatWithAggOperators(aggregationOperators, matchOps);
}

/**
 * Uses func.activeFilter_matchOp_orErrDescription_fromMultiFilter to add the matchOps for given filter(s).
 * Concats it to the given aggregationOperators and returns that or throws an error.
 * @param {Object} dataSourceDescription
 * @param {Object} options
 * @param {Object[]=} aggregationOperators
 * @returns {Object[]}
 * @throws {Error}
 */
function addFiltersToQuery(dataSourceDescription, options, aggregationOperators = []) {
    // If there are active filters we'll want to add them to the query
    if (!options.isFilterActive) {
        return aggregationOperators;
    }

    const { err, matchOps } = activeFilter_matchOp_orErrDescription_fromMultiFilter(dataSourceDescription, options.filterObj);

    if (err) {
        throw Error(`Filters not added to aggregation operators: ${err.message}`);
    }

    winston.debug('Filter is active, adding to query');

    return concatWithAggOperators(aggregationOperators, matchOps);
}
