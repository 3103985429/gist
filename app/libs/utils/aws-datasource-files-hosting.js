const winston = require('winston');
const aws = require('aws-sdk');
const fs = require('fs');
const _ = require('lodash');

const bucket = process.env.DO_S3_BUCKET;
const endpoint = new aws.Endpoint(process.env.DO_S3_ENDPOINT);
const s3 = new aws.S3({
    endpoint,
    accessKeyId: process.env.DO_ACCESS_KEY_ID,
    secretAccessKey: process.env.DO_SECRET_ACCESS_KEY,
});
const debugS3 = false;

function _uploadAncillaryFile(filePath, fileName, contentType, teamSubdomin, datasetId, callback) {
    const file = fs.createReadStream(filePath);
    const key = getAncillaryAssetKey(teamSubdomin, datasetId, fileName);
    const params = {
        Bucket: bucket,
        Key: key,
        ContentType: contentType,
        Body: file,
        ACL: 'private',
    };

    debug(`Uploading the ancillary file to S3: ${key}`);

    s3.upload(params, function (err) {
        if (err) {
            return callback(err);
        }

        callback();
    });
}

module.exports.uploadAncillaryFile = _uploadAncillaryFile;

function _copyAncillaryFile(team_subdomain, originId, destinationId, fileName, callback) {
    const key = getAncillaryAssetKey(team_subdomain, destinationId, fileName);
    const params = {
        Bucket: bucket,
        CopySource: getAncillaryAssetKey(team_subdomain, originId, fileName, bucket),
        Key: key,
    };

    debug(`Copying the ancillary file from S3: ${key}`);

    s3.copyObject(params, function (err) {
        if (err) {
            winston.error(`S3: error copying datasource: ${err.message}`);
            return callback(err);
        }

        callback();
    });
}

module.exports.copyAncillaryFile = _copyAncillaryFile;

function _uploadDataSource(filePath, newFilename, contentType, teamSubdomin, datasetId, callback) {
    const file = fs.createReadStream(filePath);
    const key = getDatasourceKey(teamSubdomin, datasetId, newFilename);

    const params = {
        Bucket: bucket,
        Key: key,
        ContentType: contentType,
        Body: file,
        ACL: 'private',
    };

    debug(`Uploading the datasource to S3: ${key}`);

    s3.upload(params, function (err) {
        if (err) {
            return callback(err);
        }

        callback();
    });
}

module.exports.uploadDataSource = _uploadDataSource;

function _copySampleDatasource(datasetToDuplicateId, fileName, datasetId, teamSubdomain, callback) {
    const key = getDatasourceKey(teamSubdomain, datasetId, fileName);
    const params = {
        Bucket: bucket,
        CopySource: getDatasourceKey('sampleteam', datasetToDuplicateId, fileName, bucket),
        Key: key,
    };

    debug(`Copying the sample datasource from S3: ${key}`);

    s3.copyObject(params, function (err) {
        if (err) {
            winston.error(`S3: error copying sample datasource: ${err.message}`);
            return callback(err);
        }

        callback();
    });
}

module.exports.copySampleDatasource = _copySampleDatasource;

function _copyDatasource(team_subdomain, fileName, copyFromId, copyToId, callback) {
    const keyFrom = getDatasourceKey(team_subdomain, copyFromId, fileName, bucket);
    const keyTo = getDatasourceKey(team_subdomain, copyToId, fileName);
    const params = {
        Bucket: bucket,
        CopySource: keyFrom,
        Key: keyTo,
    };

    debug(`Copying the datasource from S3: from ${keyFrom} to ${keyTo}`);

    s3.copyObject(params, function (err, data) {
        if (err) {
            winston.error(`S3: error copying datasource: ${err.message}`);
            return callback(err);
        }

        callback();
    });
}

module.exports.copyDatasource = _copyDatasource;

function _getDatasource(description) {
    const key = getDatasourceKey(description._team.subdomain, description.schemaId || description._id, description.fileName);

    const param = {
        Bucket: bucket,
        Key: key,
    };
    debug(`Reading the datasource from S3: ${key}`);

    return s3.getObject(param);
}

module.exports.getDatasource = _getDatasource;

function _getAncillaryFile(description) {
    const key = getAncillaryAssetKey(description._team.subdomain, description._id, description.ancillaryFile);

    const param = {
        Bucket: bucket,
        Key: key,
    };
    debug(`Reading ancillary file from S3: ${key}`);

    return s3.getObject(param);
}

module.exports.getAncillaryFile = _getAncillaryFile;

function keyExists(description, callback = _.noop) {
    return new Promise((resolve, reject) => {
        const key = getDatasourceKey(description._team.subdomain, description.schema_id || description._id, description.fileName);

        const params = {
            Bucket: bucket,
            Key: key,
        };
        debug(`Verifying the datasource from S3: ${key}`);

        s3.headObject(params, (err) => {
            if (err) {
                reject(err);
                return callback(err, false);
            }

            resolve();
            callback(null, true);
        });
    });
}

module.exports.keyExists = keyExists;

function _deleteDataset(description, callback) {
    const team_subdomain = description._team.subdomain;
    const key = getDatasetKey(team_subdomain, description.id);
    const param = {
        Bucket: bucket,
        Prefix: key,
    };

    debug(`Removing the datasource from S3: ${key}`);

    s3.listObjects(param, function (err, data) {
        if (err) {
            return callback(err);
        }

        if (!data.Contents) {
            return callback();
        }

        if (data.Contents.length === 0) {
            return callback();
        }

        const param = { Bucket: bucket };
        param.Delete = { Objects: [] };

        data.Contents.forEach(function (content) {
            param.Delete.Objects.push({ Key: content.Key });
        });

        s3.deleteObjects(param, function (err, data) {
            if (err) {
                return callback(err);
            }

            if (!data.Deleted) {
                return callback();
            }

            if (data.Deleted.length === 1000) {
                return _deleteDataset(description, callback);
            }

            callback();
        });
    });
}

module.exports.deleteDataset = _deleteDataset;

function _deleteTeam(subdomain, callback) {
    const param = {
        Bucket: bucket,
        Prefix: getTeamKey(subdomain),
    };

    s3.listObjects(param, function (err, data) {
        if (err) {
            return callback(err);
        }

        if (!data.Contents) {
            return callback();
        }

        if (data.Contents.length === 0) {
            return callback();
        }

        const param = { Bucket: bucket };
        param.Delete = { Objects: [] };

        data.Contents.forEach(function (content) {
            param.Delete.Objects.push({ Key: content.Key });
        });

        s3.deleteObjects(param, function (err, data) {
            if (err) {
                return callback(err);
            }

            if (!data.Contents) {
                return callback();
            }

            if (data.Contents.length === 1000) {
                return _deleteTeam(subdomain, callback);
            }

            callback();
        });
    });
}

module.exports.deleteTeam = _deleteTeam;

function _deleteObject(key, callback) {
    const params = {
        Bucket: bucket,
        Key: key,
    };

    s3.deleteObject(params, function (err, data) {
        if (err) {
            return callback(err);
        }

        callback(null, data);
    });
}

module.exports.deleteObject = _deleteObject;

function getDatasourceKey(subdomain, datasetId, fileName, bucket = null) {
    return getDatasetKey(subdomain, datasetId, `datasources/${fileName}`, bucket);
}

function getAncillaryAssetKey(subdomain, datasetId, fileName, bucket = null) {
    return getDatasetKey(subdomain, datasetId, `assets/ancillary/${fileName}`, bucket);
}

function getDatasetKey(subdomain, datasetId, resourceName = '', bucket = null) {
    return getTeamKey(subdomain, `datasets/${datasetId}/${resourceName}`, bucket);
}

function getTeamKey(subdomain, resourceName = '', bucket = null) {
    const prefix = bucket ? `${bucket}/` : '';

    return `${prefix}${subdomain}/${resourceName}`;
}

function debug(msg) {
    if (debugS3) {
        winston.debug(msg);
    }
}
