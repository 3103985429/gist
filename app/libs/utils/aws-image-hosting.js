const request = require('request');
var winston = require('winston');
var aws = require('aws-sdk');
var sharp = require('sharp');
var async = require('async');
const _ = require('lodash');

const { VIEW_TYPE_OBJECT_DETAILS, VIEW_TYPE_GALLERY, VIEW_TYPE_TIMELINE } = require('../../config/views.config');

const BUCKET = process.env.DO_S3_BUCKET;
const endpoint = new aws.Endpoint(process.env.DO_S3_ENDPOINT);
const S3 = new aws.S3({
    endpoint,
    accessKeyId: process.env.DO_ACCESS_KEY_ID,
    secretAccessKey: process.env.DO_SECRET_ACCESS_KEY,
    signatureVersion: 'v4',
    region: 'nyc3',
});

module.exports = {
    uploadToS3,
    signedUrlForPutObject,
    getAllIconsForTeam,
    hostImageLocatedAtRemoteURL,
    copyPageAssets,
    copyDatasetAsset,
    getAssetByKey,
};

function uploadToS3(key, response, callback) {

    var payload = {
        Bucket: BUCKET,
        Key: key,
        Body: response,
        ACL: 'public-read',
    };

    S3.upload(payload, function (err, data) {
        if (err) {
            winston.error('AWS S3 write stream error');
            return callback(err);
        }
        return callback(null, data);
    });
}

function _appendKeyToBucket(key) {
    return 'https://' + BUCKET + '.' + process.env.DO_S3_ENDPOINT + '/' + key;
}

function getImageFormat(headers) {
    const imageFormat = _.get(headers, 'content-type', '').split(';')[0].split('/')[1];

    return imageFormat === 'jpg' ? 'jpeg' : imageFormat;
}

function getImageAtUrl(remoteImageSourceURL, callback) {
    var options = {
        url: remoteImageSourceURL,
        encoding: null,
        timeout: 10000,
    };

    request.get(options, function (err, response, data) {
        if ((err && ['ENOTFOUND', 'ETIMEDOUT'].includes(err.code)) || !response) {
            winston.info(`returning url as null, since Could not read the remote image ${remoteImageSourceURL}: ${err}`);
            return callback(null, null);
        }

        if (err) {
            return callback(err, null);
        }

        const imageFormat = getImageFormat(response.headers);

        if (_.get(sharp, `format.${imageFormat}`)) {
            return callback(null, data);
        }

        return callback(null, null);
    });
}

function signedUrlForPutObject(key, fileType, callback) {
    if (fileType.indexOf('svg') >= 0) { //fix space when uploading svg error
        fileType = 'image/svg+xml';
    }

    var params = {
        Bucket: BUCKET,
        Key: key,
        ACL: 'public-read',
        ContentType: fileType,
    };

    S3.getSignedUrl('putObject', params, function (err, signedUrl) {
        callback(err, { putSignedUrl: signedUrl, publicUrl: _appendKeyToBucket(key) });
    });
}

function getAllIconsForTeam(teamSubdomain, callback) {
    S3.listObjects({
        Bucket: BUCKET,
        Prefix: teamSubdomain + '/assets/icon/',
        Marker: teamSubdomain + '/assets/icon/',
    }, function (err, data) {
        if (err) {
            callback(err);
        } else {
            var listOfUrls = [];
            var objects = data.Contents;

            for (var i = 0; i < objects.length; i++) {
                var objKey = objects[i].Key;

                var url = _appendKeyToBucket(objKey);

                listOfUrls.push(url);
            }
            callback(null, listOfUrls);
        }
    });

}

async function _proceedToStreamToHost(imageUrl, resize, folder, docPKey, callback) {
    try {
        const image = sharp(imageUrl);
        const metadata = await image.metadata();

        if (!metadata) {
            winston.info('returning url as null, since Could not read the remote image metadata');
            return callback();
        }

        const { width, hasAlpha } = metadata;
        const size = width < resize.size ? width : resize.size;
        const whiteColorRGB = { r: 255, g: 255, b: 255 };
        const background = hasAlpha ? { ...whiteColorRGB, alpha: 1 } : whiteColorRGB;
        const resizeConfig = { fit: 'contain', background };
        const resizedImage = resize.cropped ? image.resize(size, size, resizeConfig) : image.resize(size);
        const bufferedImage = await resizedImage.toBuffer();
        const keyName = `${folder}${resize.view}/${docPKey}`;

        uploadToS3(keyName, bufferedImage, callback);
    } catch (error) {
        winston.error('has err: ' + error);
        callback(error);
    }
}

function hostImageLocatedAtRemoteURL(folder, remoteImageSourceURL, overwrite, destinationFilenameSansExt, callback) {
    const recommendedSizes = [
        { size: 763, cropped: false, view: VIEW_TYPE_OBJECT_DETAILS },
        { size: 280, cropped: false, view: VIEW_TYPE_GALLERY },
        { size: 155, cropped: true, view: VIEW_TYPE_TIMELINE },
    ];

    getImageAtUrl(remoteImageSourceURL, function (err, imageBuffer) {
        if (err || !imageBuffer) {
            return callback(err);
        } else {
            async.each(recommendedSizes, function (size, cb) {
                if (overwrite) {
                    _proceedToStreamToHost(imageBuffer, size, folder, destinationFilenameSansExt, cb);
                } else {
                    const keyName = `${folder}${size.view}/${destinationFilenameSansExt}`;
                    const params = {
                        Bucket: BUCKET,
                        Key: keyName,
                    };
                    try {
                        S3.headObject(params, function (err, data) {
                            if (err || !data) {
                                winston.error('Head object err: ' + err);
                            } else {
                                winston.info(' File already uploaded and overwrite is false for object with pKey ' + destinationFilenameSansExt);
                                cb();
                            }
                        });
                    } catch (err) {

                        if (err.code === 'ENOTFOUND' || err.code === 'ETIMEDOUT') {
                            _proceedToStreamToHost(imageBuffer, size, folder, destinationFilenameSansExt, cb);
                        } else {
                            winston.error(err);
                            cb(err);
                        }
                    }
                }
            }, callback);
        }
    });

}

function copyPageAssets(args, callback) {
    var params = {
        Bucket: BUCKET,
        CopySource: BUCKET + '/' + args.subdomain + '/websites/' + args.originalWebsiteId + '/pages/' + args.originalPageId + '/' + args.imageType + '/' + args.fileName,
        Key: args.subdomain + '/websites/' + args.websiteId + '/pages/' + args.pageId + '/' + args.imageType + '/' + args.fileName,
        ACL: 'public-read',
    };

    S3.copyObject(params, function (err) {
        if (err) {
            winston.error('\' error copying page asset with params: \'', params);
            return callback(err);
        } else {
            callback();
        }
    });
}

async function copyDatasetAsset({ fromSubdomain, toSubdomain, fromId, toId, source, fileName }, callback) {
    const getUrl = ({ id, BUCKET, subdomain = fromSubdomain }) =>
        `${BUCKET ? `${BUCKET}/` : ''}${subdomain}/datasets/${id}/${source}/${fileName}`;
    const copySource = getUrl({ id: fromId, BUCKET });
    const key = getUrl({ id: toId, subdomain: toSubdomain });

    const params = {
        Bucket: BUCKET,
        CopySource: copySource,
        Key: key,
        ACL: 'public-read',
    };

    try {
        await S3.copyObject(params).promise();

        callback(null);
    } catch (e) {
        winston.error(`S3: error copying dataset asset: ${e}`);

        callback(e);
    }
}

function getAssetByKey(key, callback) {
    const params = {
        Bucket: BUCKET,
        Key: key,
    };

    S3.getObject(params, function (err, data) {
        if (err) {
            return callback(err);
        } else {
            return callback(null, data);
        }
    });
}
