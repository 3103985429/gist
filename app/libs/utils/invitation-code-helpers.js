const crypto = require('crypto');

/**
 * Returns a sha512 hash of the given string.
 * @param confirmation_code - Code to hash
 * @return
 */
module.exports.createHashWithString = confirmation_code => {
    // Create a new hash, and then update it and return the digest
    const hash = crypto.createHash('sha512');
    hash.update(confirmation_code);
    return hash.digest('hex');
};

/**
 * Generates a random short code.
 * @param codeSize - Size in bytes to generate a code
 * @return
 */
module.exports.generateRandomCode = codeSize => {
    // Get a random buffer
    const randomBuffer = crypto.randomBytes(codeSize);
    // Return the random buffer to string
    return randomBuffer.toString('hex');
};
