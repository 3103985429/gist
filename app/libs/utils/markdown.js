var fs = require('fs');
var marked = require('marked');

module.exports.getMarkdown = function (filePath) {
    var rawMarkdown = fs.readFileSync(filePath, 'utf-8');
    return rawMarkdown;
};

module.exports.parseMarkdown = function (markdown) {
    var options = {
        renderer: new marked.Renderer(),
        gfm: true,
        tables: true,
        breaks: false,
        pedantic: false,
        sanitize: true,
        smartLists: true,
        smartypants: false
    };
    marked.setOptions(options);

    var lexer = new marked.Lexer(options);
    var parsedBody = marked(markdown);
    return parsedBody;
};

module.exports.replaceCharts = function (parsed) {
    var chartRE = /((?:<p>)?&lt;chart data-source-key=&quot;(?=\S*)([a-zA-Z\d_-]*)&quot; data-chart-type=&quot;(?=\S*)([a-zA-Z\d-]*)&quot; class=&quot;chart&quot; data-revision=&quot;(\d+)&quot; \/&gt;(?:<\/p>)?)/g;
    var parsedReplaced = parsed.replace(chartRE, '<div id="$2" data-source-key="$2" data-chart-type="$3" class="$3 chart" data-revision="$4"></div>');
    return parsedReplaced;
};
