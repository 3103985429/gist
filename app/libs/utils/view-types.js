const CONTENT_ONLY_TYPE = 'contentOnly';

module.exports = {
    CONTENT_ONLY_TYPE,
};
