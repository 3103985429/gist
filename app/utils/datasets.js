const { findDataset } = require('../controllers/api/dataset/get-helpers');
const { some, includes } = require('lodash');
const DATA_SOURCE_TYPES = require('../../config/data-source-types');

module.exports.getDataset = getDataset;
module.exports.hasValidDatasetSource = hasValidDatasetSource;

/**
 * @param {String|ObjectId} id
 * @returns {Promise<Object>}
 */
async function getDataset(id) {
    const { description } = await findDataset({ id });

    return description;
}

/**
 * Determins if the datasourceDescription passed has a valid dataset source
 * determined from the list in config/data-source-types.js
 * @param {Object} datasourceDescription
 * @returns {Boolean}
 */
function hasValidDatasetSource(datasourceDescription) {
    return some(datasourceDescription, (value, key) => {
        return includes(DATA_SOURCE_TYPES, key);
    });
}
