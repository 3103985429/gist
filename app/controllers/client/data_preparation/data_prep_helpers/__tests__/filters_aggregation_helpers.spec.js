import { addSearchToQuery, addFiltersToQuery } from '../filters_aggregation_helpers';
import { hydrateDb } from '../../../../../../internals/testing/backend/utils';
import { pokemon1 } from '../../../../../../internals/testing/backend/fixtures/datasets';


describe('Search and Filter Aggregation Helpers ', () => {
    let dataSourceDescription;
    let options;
    let aggregationOperators;

    beforeEach(async () => {
        dataSourceDescription = pokemon1;
        options = {};
        aggregationOperators = [];
        await hydrateDb();
    });

    it('should not add search or filters when not activated', () => {
        // Set these options to undefined and make sure that the returned agg operators length's are 0
        options.isSearchActive = undefined;
        options.isFilterActive = undefined;

        const searchAggs = addSearchToQuery(dataSourceDescription, options, aggregationOperators);
        const filterAggs = addFiltersToQuery(dataSourceDescription, options, aggregationOperators);

        expect(searchAggs).toHaveLength(0);
        expect(filterAggs).toHaveLength(0);
    });

    it('should return a built filter with filter args passed', () => {
        // Add filterObj and expect filterAggs to not be undefined
        options.isFilterActive = true;
        options.filterObj = { 'Type 1': 'Water' };

        const filterAggs = addFiltersToQuery(dataSourceDescription, options, aggregationOperators);

        expect(filterAggs).toMatchSnapshot();
    });

    it('should return a built filter with query args passed', () => {
        // Add filter column and query, and we should get something defined back 👌🏻
        options.isSearchActive = true;
        options.searchCol = 'All Fields';
        options.searchQ = 'Bulbasaur';

        const filterAggs = addSearchToQuery(dataSourceDescription, options, aggregationOperators);

        expect(filterAggs).toMatchSnapshot();
    });

    it('should throw an error with invalid filter params passed', () => {
        // Filter active is true, and crazy invalid filterObj
        options.isFilterActive = true;
        options.filterObj = { 55: [] };

        // Function to pass to the expect and get an error thrown
        const callAddFilters = () => {
            addFiltersToQuery(dataSourceDescription, options, aggregationOperators);
        };

        expect(callAddFilters).toMatchSnapshot();
    });

    it('should preserve order of aggregationOperators', () => {
        // Add filter column and query, and we should get something defined back in the proper order 👌🏻
        options.isSearchActive = true;
        options.searchCol = 'All Fields';
        options.searchQ = 'Bulbasaur';

        // Make some fake aggregationOperators to pass
        aggregationOperators = [{ fake_operator: ':)' }];

        const filterAggs = addSearchToQuery(dataSourceDescription, options, aggregationOperators);

        expect(filterAggs[0]['fake_operator']).toBe(':)');
        expect(filterAggs.length).toBeGreaterThan(1);
    });
});
