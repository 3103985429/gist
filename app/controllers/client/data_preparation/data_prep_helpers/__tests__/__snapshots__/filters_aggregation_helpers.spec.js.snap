// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search and Filter Aggregation Helpers  should return a built filter with filter args passed 1`] = `
Array [
  Object {
    "$match": Object {
      "rowParams.Type 1": Object {
        "$options": "i",
        "$regex": "^Water$",
      },
    },
  },
  Object {
    "$group": Object {
      "_id": "$_id",
      "pKey": Object {
        "$first": "$pKey",
      },
      "rowParams": Object {
        "$first": "$rowParams",
      },
      "srcDocPKey": Object {
        "$first": "$srcDocPKey",
      },
    },
  },
]
`;

exports[`Search and Filter Aggregation Helpers  should return a built filter with query args passed 1`] = `
Array [
  Object {
    "$match": Object {
      "$or": Array [
        Object {
          "rowParams.Image URL": Object {
            "$options": "i",
            "$regex": "Bulbasaur",
          },
        },
        Object {
          "rowParams.Type 2": Object {
            "$options": "i",
            "$regex": "Bulbasaur",
          },
        },
        Object {
          "rowParams.Type 1": Object {
            "$options": "i",
            "$regex": "Bulbasaur",
          },
        },
        Object {
          "rowParams.Name": Object {
            "$options": "i",
            "$regex": "Bulbasaur",
          },
        },
      ],
    },
  },
]
`;

exports[`Search and Filter Aggregation Helpers  should throw an error with invalid filter params passed 1`] = `[Function]`;
